### ISSUE TITLE
Use of features specific to the C language

### ISSUE THREAD

**Issue by encukou** [capi-workgroup/problems#35]
Non-C languages struggle with:
- Bit fields
- Enums (which have compiler-dependent sizes)
- Macros and inline functions (which need a C compiler)
- *(edit)* varargs

Some related problems have their own issues here:
- #7  -- to parse information from headers, you generally need a C conpiler
- #27 -- type sizes are *compiler-*, not just platform-specific (although most languages have solved this for common types like `int`, by necessity)

[added 2025-01-13] Real-world issues:

- pypy: https://github.com/pypy/pypy/pull/5170

**Comment by encukou** [comment]
@davidhewitt wrote:

> How do Rust bindings handle these issues?

As for the "issues" in the OP, thus we handle them as so:

> parse C headers (which is very tricky in general)

We get humans to do this bit. To keep things easier we try to match the structure of the `include/` directory (as it is on `main`) as the `pyo3-ffi` crate. So e.g. `listobject.h` becomes the `listobject.rs` file linked to above. Naturally there is some lag between what we've synced and what's in CPython `main`.
 
> match C type sizes (e.g. sizeof(int)). Some can differ depending on the compiler, even on the same platform.

The Rust standard library has [`core::ffi`](https://doc.rust-lang.org/core/ffi/index.html) with types like `c_int`, which has so far worked for us.

> mimic C memory layout (which is also compiler-defined). Padding, bit fields, enums are particularly tricky.

Rust has `#[repr(C)]` to lay out structs in a C-compatible way and can do C-compatible enums. Bitfields are extremely awkward; we have to use integer fields of the correct total size and implement offset operations to do the reads. `bindgen` has helped us get this correct in the past.

> do ... something? ... with C macros and inline functions.

We've reproduced most of these as `#[inline] unsafe fn`, e.g. here's the implementation for [`Py_TYPE`](https://github.com/PyO3/pyo3/blob/0f0024026beb5dc00b4d8a677915d35418d70079/pyo3-ffi/src/object.rs#L70-L73). This obviously isn't perfect; macros which aren't functions such as `PyDateTime_TimeZone_UTC` will still get mirrored as functions, i.e. `PyDateTime_TimeZone_UTC()`.

**Comment by steve-s** [comment]
I think that providing some convenience helpers specific for C is fine, but any C specific should be left out from the ABI, or have alternative non-C specific variants in the ABI. For example, `PyErr_FromFormat` is extremely useful if you are in C. If you are developing rust language binding, then you probably want to provide rust-style formatting helper that fits the rust way of doing such things. The ABI should have you covered without having to do some cumbersome C interop.

**Comment by vstinner** [comment]
@steve-s:
> Edit: now I see you don't list varargs, but I'd add them to the list too?

What is the problem with variadic arguments? Are there programming languages which cannot use them and are used to write Python extensions?

Are you saying that we should **not** add functions using variadic arguments in their defintion, like ``PyObject* Py_BuildValue(const char *format, ...)``? Or are you saying that we should **not** add functions using ``vargs: va_list`` (``<stdarg.h>``) in their definition, like ``PyObject* PyUnicode_FromFormatV(const char *format, va_list vargs)``)?

How do you implement a function with a variable number of arguments in these languages? For example, how do you bind ``Py_BuildValue()``, ``PyTuple_Pack()``, or ``PyUnicode_FromFormat()`` in these languages?

I looked at PyO3 (Rust), it seems to happily bind these functions:

* [Py_BuildValue()](https://github.com/PyO3/pyo3/blob/e67b283b25eb120953720a0fcc62c2943f1abbf1/pyo3-ffi/src/modsupport.rs#L29-L30)
* [PyTuple_Pack()](https://github.com/PyO3/pyo3/blob/e67b283b25eb120953720a0fcc62c2943f1abbf1/pyo3-ffi/src/tupleobject.rs#L38-L39)
* [PyUnicode_FromFormat()](https://github.com/PyO3/pyo3/blob/e67b283b25eb120953720a0fcc62c2943f1abbf1/pyo3-ffi/src/unicodeobject.rs#L86-L87).

PyO3 also has PyUnicode_FromFormatV(), but [its implementation is commented](https://github.com/PyO3/pyo3/blob/e67b283b25eb120953720a0fcc62c2943f1abbf1/pyo3-ffi/src/unicodeobject.rs#L84C38-L85).

**Comment by vstinner** [comment]
The question of variadic arguments arised in my issue proposing to add ``int PySys_Audit(const char *event, const char *argFormat, ...)`` to the limited C API: https://github.com/python/cpython/pull/108571

**Comment by encukou** [comment]
I don't think we should avoid functions like `Py_BuildValue`. But they should not be the *only* way to do something.

For `Py_BuildValue` you can use `PyTuple_New`+`PyTuple_SetItem`, or build from a list, or (ideally in the future) build from an array of `PyObject*`.

For `PyUnicode_FromFormatV` you can prepare the `char*` using any other formatting API, and call `PyUnicode_FromStringAndSize`.

For C, the varargs variants might still be the right way to do things. But alternatives should (and do) exist.

**Comment by vstinner** [comment]
> For Py_BuildValue you can use PyTuple_New+PyTuple_SetItem, or build from a list, or (ideally in the future) build from an array of PyObject*.

Note: IMO this API is bad since it creates an uninitialized Python tuple object, see: https://github.com/capi-workgroup/problems/issues/56

> For PyUnicode_FromFormatV you can prepare the char* using any other formatting API, and call PyUnicode_FromStringAndSize.

Yep, ``PyUnicode_FromStringAndSize()`` is a nice API in term of API stability and not leaking implementation details.

**Comment by vstinner** [comment]
What said @encukou for PySys_Audit() now makes sense to me: if we add an API with variadic arguments, we should offer a similar API without it. PySys_Audit() builds a tuple: the second flavor would be an API which already takes a tuple.

So people would can use variadic arguments just use it (it's a convenient API), and there is a backup plan for other people.

Thanks @cavokz, I didn't want to add a new function and have to maintain it to solve an hypothetical problem. I didn't know that Go doesn't support variadic arguments and that Python can be used with Go, two things that I learned today 👍

**Comment by steve-s** [comment]
To clarify, what I meant:
* ABI should provide all the necessary functionality without the need for varargs (we seem to be in agreement on this)
* I am not sure how stable and well defined are varagrs w.r.t. ABI (they probably are in practice, is their ABI standardized?, is it safe to rely on that)
* Note that in HPy, the ABI is varargs free, but HPy provides helper vargars functions that are not part of ABI and are implemented using the ABI. The helper functions are compiled into the extension. The idea is that language bindings can piggy back on the C API (if they have good C interop), but they can also just build on top of the ABI and provide a facade for whatever interface is natural/idiomatic for given language.

**Comment by encukou** [comment]
> I am not sure how stable and well defined are varagrs w.r.t. ABI (they probably are in practice, is their ABI standardized?, is it safe to rely on that)

AFAIK, C standards say nothing about ABI. This kind of stuff is left to platforms/distributors, see the [disclaimer in the docs](https://docs.python.org/3/c-api/stable.html#platform-considerations)

> Note that in HPy, the ABI is varargs free, but HPy provides helper vargars functions that are not part of ABI and are implemented using the ABI. The helper functions are compiled into the extension. The idea is that language bindings can piggy back on the C API (if they have good C interop), but they can also just build on top of the ABI and provide a facade for whatever interface is natural/idiomatic for given language.

That's the general idea of the direction I want to go in: a very portable base, and optional C-specific niceties.
(The  `PySys_Audit` case is uniquely tricky though. If no hooks are registered, `PySys_Audit` avoids wrapping arguments in `PyObject*s`. And AFAIK, for security reasons we can't tell the caller whether hooks are registered, so users of the tuple API can't avoid allocating the tuple.)

**Comment by encukou** [comment]
Issues for proposed guidelines:
- https://github.com/capi-workgroup/api-evolution/issues/10
- https://github.com/capi-workgroup/api-evolution/issues/11
- https://github.com/capi-workgroup/api-evolution/issues/12
- https://github.com/capi-workgroup/api-evolution/issues/18

**Comment by pitrou** [comment]
> * Enums (which have compiler-dependent sizes)

I see this mentioned in many places, but the enum size is actually mandated by the platform ABI, which the compiler is supposed to comply with.

For example the SysV ABI for x86-64 (used on GNU/Linux) mandates that enum have size 4 and alignment 4, except when some enum values wouldn't fit (which is unlikely):
https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf

Similarly, Microsoft's x86-64 ABI has a similar provision and doesn't even seem to make rooms for enums larger than 32 bits (_"enums are constant integers and are treated as 32-bit integers"_):
https://learn.microsoft.com/en-us/cpp/build/x64-software-conventions?view=msvc-170#scalar-types

That said, I agree that having to _think_ about enum size when using a non-C/C++ language is annoying in itself.

**Comment by encukou** [comment]
See also this pypy issue: https://github.com/pypy/pypy/pull/5170
